marlowe/test/Spec/Marlowe/Semantics/Functions.hs

```diff
@@ -0,0 +1,961 @@
+        x = AvailableMoney account token
+      in
+        case (account, token) `AM.lookup` accounts of
+          Nothing -> eval x == 0
+          Just x' -> eval x == x'
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.Constant`.
+checkConstant :: Property
+checkConstant =
+  checkValue (const . const $ arbitrary) $ \eval _ _ _ x ->
+    eval (Constant x) == x
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.NegValue`.
+checkNegValue :: Property
+checkNegValue =
+  checkValue (const . const $ arbitrary) $ \eval _ _ _ x ->
+    eval (NegValue x) == - eval x
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.AddValue`.
+checkAddValue :: Property
+checkAddValue =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval _ _ _ (x, y) ->
+      eval (AddValue x y) == eval x + eval y
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.SubValue`.
+checkSubValue :: Property
+checkSubValue =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval _ _ _ (x, y) ->
+      eval (SubValue x y) == eval x - eval y
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.MulValue`.
+checkMulValue :: Property
+checkMulValue =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval _ _ _ (x, y) ->
+      eval (MulValue x y) == eval x * eval y
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.DivValue` for 0/0.
+checkDivValueNumeratorDenominatorZero :: Assertion
+checkDivValueNumeratorDenominatorZero =
+  assertBool "DivValue 0 0 = 0"
+    $ evalValue undefined undefined (DivValue (Constant 0) (Constant 0)) == 0
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.DivValue` for zero numerator.
+checkDivValueNumeratorZero :: Property
+checkDivValueNumeratorZero =
+  checkValue (const . const $ arbitrary) $ \eval _ _ _ x ->
+    eval (DivValue (Constant 0) x) == 0
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.DivValue` for zero denominator.
+checkDivValueDenominatorZero :: Property
+checkDivValueDenominatorZero =
+  checkValue (const . const $ arbitrary) $ \eval _ _ _ x ->
+    eval (DivValue x (Constant 0)) == 0
+
+
+-- | Test ` Language.Marlowe.Core.V1.Semantics.Types.DivValue . Language.Marlowe.Core.V1.Semantics.Types.MulValue`.
+checkDivValueMultiple :: Property
+checkDivValueMultiple =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval _ _ _ (x, n) ->
+      eval (DivValue (MulValue x n) n) == eval x || eval n == 0
+
+
+-- | Test rounding of `Language.Marlowe.Core.V1.Semantics.Types.DivValue`.
+checkDivValueRounding :: Property
+checkDivValueRounding =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval _ _ _ (x, y) ->
+      eval (DivValue x y) == eval x `truncatedDivide` eval y || eval y == 0
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.ChoiceValue`.
+checkChoiceValue :: Bool      -- ^ Whether the choices is present in the state.
```

Comment:
```suggestion
checkChoiceValue :: Bool      -- ^ Whether the choice should be taken from the state.
```
There is a grammar typo, but I also wonder whether "should be taken from state" is more accurate